xmtp_mls/src/subscriptions.rs

```diff
@@ -464,81 +443,82 @@ mod tests {
             .await
             .unwrap();
 
-        tokio::time::sleep(std::time::Duration::from_millis(100)).await;
-
         let messages: Arc<Mutex<Vec<StoredGroupMessage>>> = Arc::new(Mutex::new(Vec::new()));
         let messages_clone = messages.clone();
-        let stream =
+        let notify = Arc::new(tokio::sync::Notify::new());
+        let notify_pointer = notify.clone();
+        let handle =
             Client::<GrpcClient>::stream_all_messages_with_callback(caro.clone(), move |message| {
-                let text = String::from_utf8(message.decrypted_message_bytes.clone())
-                    .unwrap_or("<not UTF8>".to_string());
-                println!("Received: {}", text);
+                notify_pointer.notify_one();
                 (*messages_clone.lock().unwrap()).push(message);
-            })
-            .await
-            .unwrap();
-
-        tokio::time::sleep(std::time::Duration::from_millis(50)).await;
+            });
+        let handle = handle.wait_for_ready().await;
 
         alix_group
             .send_message("first".as_bytes(), &alix)
             .await
             .unwrap();
-        tokio::time::sleep(std::time::Duration::from_millis(100)).await;
-
+        notify.notified().await;
+        
         let bo_group = bo
             .create_group(None, GroupMetadataOptions::default())
             .unwrap();
         bo_group
             .add_members_by_inbox_id(&bo, vec![caro.inbox_id()])
             .await
             .unwrap();
-        tokio::time::sleep(std::time::Duration::from_millis(300)).await;
 
         bo_group
             .send_message("second".as_bytes(), &bo)
             .await
             .unwrap();
-        tokio::time::sleep(std::time::Duration::from_millis(100)).await;
+        notify.notified().await;
 
         alix_group
             .send_message("third".as_bytes(), &alix)
             .await
             .unwrap();
-        tokio::time::sleep(std::time::Duration::from_millis(100)).await;
-
+        notify.notified().await;
+        
         let alix_group_2 = alix
             .create_group(None, GroupMetadataOptions::default())
             .unwrap();
         alix_group_2
             .add_members_by_inbox_id(&alix, vec![caro.inbox_id()])
             .await
             .unwrap();
-        tokio::time::sleep(std::time::Duration::from_millis(300)).await;
+        
+        // TODO:
+        // theres missed messages here IF:
```

Comment:
Got stuck here debugging for a bit with lots of improvised logs before figuring out what the issue is:

If a message is sent right after a new group is created, we can run into an issue where we actually miss messages on the stream. The message will be sent/published while the old stream is active, and then once the new stream is created it mysteriously disappears. At least, that's what i believe the issue to be. 

I think this is because of the round-trip to the gRPC and back. I reproduced this issue with the code on `main` before refactor, too, so it's not a new issue but was masked because of the `std::time::sleeps`.

Not sure if there is a good way to avoid this, or we just have to live with it. Maybe some kind of update function into gRPC instead of re-initalizing the stream? Would love your input here @richardhuaaa or @neekolas 

Maybe not a huge deal though. I'm assuming sdks have functions for filling in missing messages?

Reply:
While typing this out I realized a possible solution which i did [here](https://github.com/xmtp/libxmtp/pull/835/commits/2bae690cf1625297ebd91d5c258c182681fe6f52)

One of the problems is that the condition happens when the message is sent _just as_ the new stream is initializing. The new stream initializes and dropped while the message is still present.

Also created a test to better represent this case